packages/workbox-background-sync/lib/Queue.mjs

```diff
@@ -0,0 +1,208 @@
+/*
+ Copyright 2017 Google Inc. All Rights Reserved.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import {_private} from 'workbox-core';
+import {TAG_PREFIX, MAX_RETENTION_TIME} from './constants.mjs';
+import QueueStore from './QueueStore.mjs';
+import StorableRequest from './StorableRequest.mjs';
+import '../_version.mjs';
+
+
+const {WorkboxError} = _private;
+const queueNames = new Set();
+
+
+/**
+ * A class to manage storing failed requests in IndexedDB and retrying them
+ * later. All parts of the storing and replaying process are observable via
+ * callbacks.
+ */
+export default class Queue {
+  /**
+   * Creates an instance of Queue with the given options
+   *
+   * @param {string} name The unique name for this queue. This name must be
+   *     unique as it's used to register sync events and store requests
+   *     in IndexedDB specific to this instance. An error will be thrown if
+   *     a duplicate name is detected.
+   * @param {Object} [param2]
+   * @param {Object} [param2.callbacks] Callbacks to observe the lifecycle of
+   *     queued requests. Use these to respond to or modify the requests
+   *     during the replay process.
+   * @param {function(StorableRequest):undefined}
```

Comment:
What is the syntax for `:undefined`? The parameter is marked as optional given the square brackets.

Reply:
This means the function returns `undefined`. The syntax is described here: https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System

But that brings up a good point that this is closure compiler jsdoc notation, and I'm not sure if what we use to generate docs will support this. I don't think non-CC-jsdoc syntax has a way to specify what params a function takes, what its `this` binding is, or what it returns (which is too bad, IMO).